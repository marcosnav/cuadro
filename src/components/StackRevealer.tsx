import React, { FC, useCallback, useLayoutEffect, useState } from 'react';
import * as S from './styled';

interface IProps {
  children: React.ReactNode[];
  show?: boolean;
  direction?: 'UP'|'DOWN';
}

const childrenStack = (children: React.ReactNode[], direction: 'UP'|'DOWN' = 'UP', showTracker: number) => {
  const stack = children.map((child, index) => {
    return (
      <S.Revealer
        key={'stackchild-' + index}
        direction={direction}
        show={index <= showTracker}
      >
        {child}
      </S.Revealer>
    );
  });
  return stack;
};

const StackRevealer: FC<IProps> = ({ children, direction, show }) => {
  const [showTracker, setTrackIndex] = useState(-1);

  const updateStackReveal = useCallback(() => {
    setTimeout(() => {
      setTrackIndex(showTracker + 1);
    }, 400);
  }, [showTracker]);

  useLayoutEffect(() => {
    if (show && children.length > showTracker) {
      updateStackReveal();
    }
  }, [show, showTracker, children, updateStackReveal]);

  return (
    <div>
      {childrenStack(children, direction, showTracker)}
    </div>
  );
};

StackRevealer.defaultProps = {
  show: false,
};

export default StackRevealer;
