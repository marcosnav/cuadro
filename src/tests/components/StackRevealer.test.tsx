import { mount } from 'enzyme';
import React from 'react';
import { act } from 'react-dom/test-utils';
import StackRevealer from './../../components/StackRevealer';

describe('<StackRevealer/>', () => {
  const component = mount(<StackRevealer show={false}>{[(<div key='1'/>), (<div key='2'/>)]}</StackRevealer>);

  test('renders properly', () => {
    const revealers = component.find('Revealer');
    expect(revealers.length).toBe(2);
    expect(revealers.first().props().show).toBe(false);
    expect(revealers.first().props().direction).toBe('UP');
    expect(component).toMatchSnapshot();
  });

  test('show: true', () => {
    jest.useFakeTimers();

    act(() => {
      component.setProps({ show: true });
    });

    // Initial Render
    expect(setTimeout).toHaveBeenCalledTimes(1);
    expect(setTimeout).toHaveBeenLastCalledWith(expect.any(Function), 400);
    component.update();
    let revealers = component.find('Revealer');
    expect(revealers.first().props().show).toBe(false);
    expect(revealers.last().props().show).toBe(false);

    // First timer
    act(() => {
      jest.advanceTimersByTime(400);
    });
    expect(setTimeout).toHaveBeenCalledTimes(2);
    expect(setTimeout).toHaveBeenLastCalledWith(expect.any(Function), 400);
    component.update();
    revealers = component.find('Revealer');
    expect(revealers.first().props().show).toBe(true);
    expect(revealers.last().props().show).toBe(false);

    // Second timer
    act(() => {
      jest.advanceTimersByTime(400);
    });
    expect(setTimeout).toHaveBeenCalledTimes(3);
    expect(setTimeout).toHaveBeenLastCalledWith(expect.any(Function), 400);
    component.update();
    revealers = component.find('Revealer');
    expect(revealers.first().props().show).toBe(true);
    expect(revealers.last().props().show).toBe(true);
  });
});
