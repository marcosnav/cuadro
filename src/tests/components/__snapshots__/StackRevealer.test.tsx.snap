// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StackRevealer/> renders properly 1`] = `
<div>
  <Revealer
    direction="UP"
    show={false}
  >
    <div
      className="sc-cSHVUG dFRzVC"
      direction="UP"
    >
      <div />
    </div>
  </Revealer>
  <Revealer
    direction="UP"
    show={false}
  >
    <div
      className="sc-cSHVUG dFRzVC"
      direction="UP"
    >
      <div />
    </div>
  </Revealer>
</div>
`;
